### PR TITLE
CRM-19948: Store Attachment uploader information 

### PR DESCRIPTION
**Pull request by monishdeb**
This PR changes the Attachment API in order for it to save the current logged in contact ID as the uploader of the file.

The ID is stored in the `created_id` field, which was added to the File entity. On `Attachment.create`, only when adding a new file (i.e. File.id is empty), we set the `created_id` as logged in contact ID.

The `created_id` field is not required and, on delete, it's set to null. This was done because:
- There might be existent files on the database
- A few other places use the File BAO directly and we don't want/need to update all them right now
- When deleting a contact, we still might want to keep the file.

Note: I couldn't find a way to test the upgrader, so I'm not really sure if it's working. I tried to play with https://github.com/civicrm/civicrm-upgrade-test but didn't have any success

---

 * [CRM-19948: Store the Attachment uploader information](https://issues.civicrm.org/jira/browse/CRM-19948)

### CODE REVIEW

**Comment by monishdeb** [comment]
ping @colemanw @totten @eileenmcnaughton

File: api/v3/Attachment.php [hunk 1]
```diff
@@ -147,6 +147,7 @@ function civicrm_api3_attachment_create($params) {
   $fileDao->copyValues($file);
   if (!$id) {
     $fileDao->uri = CRM_Utils_File::makeFileName($name);
+    $fileDao->created_id = CRM_Core_Session::getLoggedInContactID();
```
Comment (eileenmcnaughton) [inline api/v3/Attachment.php]: I think it's better to do this on the spec function - ie 
    'api.default' => 'user_contact_id',

**Comment by eileenmcnaughton** [comment]
@monishdeb I understand this PR is a reviewer's commit so if you have tested it I'm Ok with it - although I made one stylistic comment.

I think there is a pretty strong precedent for adding this sort of column and while adding columns to big tables should be done cautiously due to the length of time the upgrade script can take  (we can't impose 1-2 hour outages every month) my feeling is that the file table is not likely to be massive even for large sites.

Regarding the upgrader test - you mean you couldn't just run the upgrade locally?

File: api/v3/Attachment.php (hunk 1 above)
Comment (monishdeb) [reply api/v3/Attachment.php]: yeah right .. how could I overlooked it :( Will do it

**Comment by monishdeb** [comment]
Thanks, @eileenmcnaughton for your feedback. I have made the additional fix, also earlier I tested the patch over a multilingual upgrade on my local. And the test build covers the monolingual upgrade, so I think the PR is safe to merge.

**Comment by colemanw** [comment]
@monishdeb I'm surprised the `AllCoreTables.data.php` file didn't get updated by codeGen to use square bracket syntax for arrays. Did you have to do any manual updating of the file or were there any issues with codeGen?

**Comment by colemanw** [comment]
@monishdeb I'm still very confused as to how the arrays managed to stay in the old syntax after being passed through codeGen the first time you did this, and why they are suddenly coming out in the correct syntax now, even though we haven't done any updates to codeGen.

**Comment by monishdeb** [comment]
@colemanw that's because the CodeGen will simply ignore if that file is already present (see [here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/CodeGen/Main.php#L97) where `CRM_Core_CodeGen_Reflection::needsUpdate()` return TRUE if that files is not present and eventually hits the gencode). Reason why we were not getting the correct syntax. I have then deleted the `CRM/Core/DAO/AllCoreTables.data.php` and then executed the `bin/regen.sh` which corrected syntax. (Another alternative is to run `sh bin/regen.sh -g`)

**Comment by totten** [comment]
>> I'm still very confused as to how the arrays managed to stay in the old syntax after being passed through codeGen

> that's because the CodeGen will simply ignore if that file is already present 

This is incorrect. The [Reflection::needsUpdate()](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/CodeGen/Reflection.php#L19-L32) check says that:

* If the file does **not** exist, then it **must** be updated.
* If the file **does** exist, then generate a quick preview (`$this->getRaw()`) and compare that to the actual code (using an  an approximate equivalence test, `isApproxPhpMatch(...)`).

The problem here is that the equivalence-test (`isApproxPhpMatch()`) is too clever, owing to https://github.com/civicrm/civicrm-core/commit/c3fc2621677b114dd8b622ef88a3ceec0ffbcf59#diff-ec73ce9b7d53e38552af8595f882ae9bR77 -- note how line 77 means that both `$expected` and `$actual` will be filtered through `ArraySyntaxConverter` before comparing them. In effect, that line says: "two PHP files are equivalent if all the code is the same, *notwithstanding array-syntax variations*". 

I suppose there is a legitimate argument that the two files are truly equivalent (because array-syntax differences are trivial). But I can't think of a reason why we'd want to treat them as equivalent for `GenCode` purposes. We'd rather err towards having the current `*.php` match the current `*.tpl`. Just remove the call to `ArraySyntaxConverter`, run `GenCode`, and make sure you commit the current revisions of any generated files.

**Comment by colemanw** [comment]
@totten the reason for c3fc262 is that our old php beautifier is incapable of processing files with short array syntax. Also, the smarty function `{print_array}` does not seem capable of outputting short syntax. So when generating DAOs, the existing files will have short syntax and the generated will have old syntax. Hence the necessity to convert them before comparing as strings.

**Comment by monishdeb** [comment]
@colemanw @totten I have rebased this PR and shifted the upgrade code to 5.2alpha1. Please have a look!

File: api/v3/Attachment.php [hunk 2]
```diff
@@ -493,6 +495,12 @@ function _civicrm_api3_attachment_getfields() {
     'description' => 'File content (not searchable, not returned by default)',
     'type' => CRM_Utils_Type::T_STRING,
   );
+  $spec['created_id'] = array(
+    'title' => 'Created By Contact ID',
+    'type' => CRM_Utils_Type::T_INT,
+    'description' => 'FK to civicrm_contact, who uploaded this file',
+    'api.default' => 'user_contact_id',
```
Comment (colemanw) [inline api/v3/Attachment.php]: I wonder if the File BAO is a more appropriate place to set this default value than the API? Setting it here will only affect apiV3.

File: CRM/Upgrade/Incremental/php/FiveTwo.php [hunk 3]
```diff
@@ -61,6 +61,32 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
     // }
   }
 
+  /**
+   * Upgrade function.
+   *
+   * @param string $rev
+   */
+  public function upgrade_5_2_alpha1($rev) {
```
Comment (monishdeb) [inline CRM/Upgrade/Incremental/php/FiveTwo.php]: @eileenmcnaughton like I was saying if we need to set it to ```upgrade_5_2_alpha1```

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveTwo.php]: oh - will watch & see if @totten / @colemanw merge this :-)

File: CRM/Core/BAO/File.php [hunk 4]
```diff
@@ -40,6 +40,27 @@ class CRM_Core_BAO_File extends CRM_Core_DAO_File {
...
+   * @param array $params
+   *   (reference ) an assoc array of name/value pairs.
+   *
+   * @return CRM_Core_BAO_File
+   */
+  public static function create($params) {
+    $fileDAO = new CRM_Core_DAO_File();
+    $fileDAO->copyValues($params);
+
+    if (empty($params['created_id'])) {
```
Comment (colemanw) [inline CRM/Core/BAO/File.php]: I think this ought to also be conditional if there is no `$params['id']` to avoid overwriting this value on update.

**Comment by monishdeb** [comment]
@colemanw @eileenmcnaughton can you please review & merge this PR?

File: CRM/Core/BAO/File.php [hunk 5]
```diff
@@ -40,6 +40,27 @@ class CRM_Core_BAO_File extends CRM_Core_DAO_File {
...
+   */
+  public static function create($params) {
+    $fileDAO = new CRM_Core_DAO_File();
+    $fileDAO->copyValues($params);
+
+    if (empty($params['id']) && empty($params['created_id'])) {
+      $fileDAO->created_id = CRM_Core_Session::getLoggedInContactID();
+    }
+
+    $fileDAO->save();
```
Comment (colemanw) [inline CRM/Core/BAO/File.php]: I think this BAO::create() function needs to call pre/post hooks.

**Review by colemanw — CHANGES_REQUESTED** [review]
I think this is good except for missing hooks.

File: CRM/Core/BAO/File.php (hunk 5 above)
Comment (monishdeb) [reply CRM/Core/BAO/File.php]: oops missed it, thanks for pointing that out.

 Added the hook calls. Can you please check now?

**Review by colemanw — APPROVED** [review]

**Comment by eileenmcnaughton** [comment]
@colemanw what does approved mean? How does that fit with 'good to merge'

**Comment by colemanw** [comment]
@eileenmcnaughton I had been about to merge it but then waned to verify with @totten that the upgrade code was going into the correct place. Got that question answered so now merging.

**Pull request merged**